### PR TITLE
Add some derived benchmarks

### PR DESCRIPTION
**Pull request by adithyaov**
- span
- spanBy
- spanByRolling

### CODE REVIEW

File: benchmark/Streamly/Benchmark/Data/Parser/ParserD.hs [hunk 1]
```diff
@@ -229,6 +262,13 @@ o_1_space_serial value =
     , benchIOSink value "teeFst (all,any)" $ teeFstAllAny value
     , benchIOSink value "shortest (all,any)" $ shortestAllAny value
     , benchIOSink value "longest (all,any)" $ longestAllAny value
+    ] ++ o_1_space_serial_derived value
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Parser/ParserD.hs:265]: Add it in the main function instead.

File: benchmark/Streamly/Benchmark/Data/Parser/ParserD.hs [hunk 2]
```diff
@@ -165,6 +165,39 @@ longestAllAny value =
...
+-------------------------------------------------------------------------------
+
+{-# INLINE span #-}
+span :: MonadThrow m => Int -> SerialT m Int -> m ()
+span value = IP.parseD (span_ (<= (value `div` 2)) FL.drain FL.drain)
+
+    where
+
+    span_ p f1 f2 =
+        PR.splitWith (\_ _ -> ()) (PR.takeWhile p f1) (PR.fromFold f2)
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Parser/ParserD.hs:179]: The code would be clearer if we define the two parsers p1 and p2 and then use `splitWith f p1 p2`.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

**Review by harendra-kumar — COMMENTED** [review]
Is the performance same using parsers?

**Comment by adithyaov** [comment]
```
benchmarked Data.Fold/o-1-space/composition/span
time                 49.08 μs   (48.74 μs .. 49.35 μs)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 48.89 μs   (48.57 μs .. 49.64 μs)
std dev              2.477 μs   (1.007 μs .. 4.534 μs)
variance introduced by outliers: 49% (moderately inflated)

benchmarked Data.Fold/o-1-space/composition/spanBy
time                 49.48 μs   (49.32 μs .. 49.62 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 50.80 μs   (50.18 μs .. 51.86 μs)
std dev              4.032 μs   (2.979 μs .. 5.497 μs)
variance introduced by outliers: 70% (severely inflated)

benchmarked Data.Fold/o-1-space/composition/spanByRolling
time                 49.70 μs   (49.48 μs .. 49.92 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 49.76 μs   (49.57 μs .. 49.98 μs)
std dev              998.3 ns   (873.7 ns .. 1.171 μs)
variance introduced by outliers: 13% (moderately inflated)
```

```
benchmarked Data.Parser.ParserD/o-1-space/span
time                 1.285 ms   (1.268 ms .. 1.302 ms)
                     0.999 R²   (0.998 R² .. 0.999 R²)
mean                 1.176 ms   (1.132 ms .. 1.209 ms)
std dev              121.8 μs   (85.25 μs .. 164.8 μs)
variance introduced by outliers: 62% (severely inflated)

benchmarked Data.Parser.ParserD/o-1-space/spanBy
time                 1.316 ms   (1.297 ms .. 1.332 ms)
                     0.998 R²   (0.996 R² .. 0.999 R²)
mean                 1.187 ms   (1.127 ms .. 1.227 ms)
std dev              155.3 μs   (112.6 μs .. 210.8 μs)
variance introduced by outliers: 72% (severely inflated)

benchmarked Data.Parser.ParserD/o-1-space/spanByRolling
time                 1.322 ms   (1.303 ms .. 1.339 ms)
                     0.999 R²   (0.998 R² .. 0.999 R²)
mean                 1.190 ms   (1.131 ms .. 1.230 ms)
std dev              155.5 μs   (115.4 μs .. 206.9 μs)
variance introduced by outliers: 72% (severely inflated)
```

```
benchmarked Data.Parser.ParserD/o-1-space/splitWith
time                 1.285 ms   (1.266 ms .. 1.302 ms)
                     0.999 R²   (0.998 R² .. 0.999 R²)
mean                 1.183 ms   (1.137 ms .. 1.217 ms)
std dev              124.9 μs   (88.67 μs .. 175.5 μs)
variance introduced by outliers: 62% (severely inflated)
```

**Comment by adithyaov** [comment]
```
benchmarked Data.Parser.ParserD/o-1-space/span
time                 69.17 μs   (68.90 μs .. 69.46 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 69.35 μs   (69.02 μs .. 69.74 μs)
std dev              1.774 μs   (1.462 μs .. 2.250 μs)
variance introduced by outliers: 19% (moderately inflated)

benchmarked Data.Parser.ParserD/o-1-space/spanBy
time                 69.20 μs   (68.91 μs .. 69.46 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 70.26 μs   (69.97 μs .. 71.09 μs)
std dev              2.186 μs   (1.183 μs .. 4.527 μs)
variance introduced by outliers: 25% (moderately inflated)

benchmarked Data.Parser.ParserD/o-1-space/spanByRolling
time                 68.37 μs   (68.18 μs .. 68.59 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 69.39 μs   (69.11 μs .. 69.69 μs)
std dev              1.421 μs   (1.229 μs .. 1.666 μs)
variance introduced by outliers: 12% (moderately inflated)
```

**Review by harendra-kumar — APPROVED** [review]

**Pull request merged**